extensions/data-transfer/portability-data-transfer-imgur/src/main/java/org/datatransferproject/datatransfer/imgur/photos/ImgurPhotosExporter.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright 2019 The Data Transfer Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.datatransferproject.datatransfer.imgur.photos;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.datatransferproject.api.launcher.Monitor;
+import org.datatransferproject.spi.transfer.provider.ExportResult;
+import org.datatransferproject.spi.transfer.provider.Exporter;
+import org.datatransferproject.types.common.ExportInformation;
+
+import org.datatransferproject.types.transfer.auth.TokensAndUrlAuthData;
+import org.datatransferproject.types.common.models.photos.PhotoAlbum;
+import org.datatransferproject.types.common.models.photos.PhotoModel;
+import org.datatransferproject.types.common.models.photos.PhotosContainerResource;
+
+import java.io.IOException;
+import java.util.*;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+
+/** Exports Imgur albums and photos using Imgur API */
+public class ImgurPhotosExporter
+    implements Exporter<TokensAndUrlAuthData, PhotosContainerResource> {
+  private final String albumPhotosUrlTemplate;
+  private final String allPhotosUrlTemplate;
+  private final String albumsUrlTemplate;
+  private final String defaultAlbumId;
+
+  private final OkHttpClient client;
+  private final ObjectMapper objectMapper;
+  private Monitor monitor;
+
+  public ImgurPhotosExporter(Monitor monitor, String baseUrl) {
+    albumPhotosUrlTemplate = baseUrl + "/album/%s/images";
+    albumsUrlTemplate = baseUrl + "/account/me/albums/%s";
+    allPhotosUrlTemplate = baseUrl + "/account/me/images/%s";
+    defaultAlbumId = "0";
+    this.client = new OkHttpClient.Builder().build();
+    this.objectMapper = new ObjectMapper();
+    this.monitor = monitor;
+  }
+
+  /**
+   * Exports albums and photos. Gets albums first, then photos which are contained in albums and
+   * non-album photos
+   *
+   * @param jobId the job id
+   * @param authData authentication data for the operation
+   * @param exportInformation info about what data to export, see {@link ExportInformation} for more
+   */
+  @Override
+  public ExportResult<PhotosContainerResource> export(
+      UUID jobId, TokensAndUrlAuthData authData, Optional<ExportInformation> exportInformation)
```

Comment:
We're not making use of the ExportInformation here since we're just getting all of the content in one go. 

Lets try and split it up, some ideas: 
- common pattern is you either retrieve the albums list or the photos list at each call based on the exportInformation. If its empty, it means we haven't gotten anything yet and we should export the list of albums, if there is a an album we should retrieve the data for it (See Flickr Exporter for an example of this)
- You mentioned the API doesn't allow to get non-album photos separately - when you retrieve the photos do they include which album they belong in? The other thing we can do is retrieve the list of albums, then all photos and add them to albums as we process each photo.  But again, so we dont request all of them at the same time, we include the pagination token in the ExportResult and when we enter this function, we pick up from the page that we left off at. 

Let me know if that makes sense, also adding chuy@ to take a look as well since its been a while since I've written or reviewed transfer extensions.

Reply:
I refactored it and now exporter uses ExportInformation. Unfortunately retrieved photos don't include album (here is it's [data model](https://api.imgur.com/models/image) )
Exporter performs these steps:
- At first gets albums (with pagination) and saves their ids to the ExportInformation.  
- Retrieves photos for each album on the subsequent export calls (request doesn't support pagination), saves their ids to the Set.
- Exports non-album photos: retrieves all photos (with pagination) and excludes album photos saved to the Set.